### PR TITLE
Add Postgres to Mathesar docker image

### LINKED ISSUE #3072: Add a Docker image containing the Mathesar server along with a Postgres server
## Problem
The current docker image requires the user to already have a Postgres server running on their machine. This is a hassle for someone looking to try out Matheasar quickly.

## Proposed solution
Provide a docker image containing a Postgres server along with the Mathesar server

### PR DESCRIPTION
**Pull request by silentninja**
Fixes #3072
The PR adds a Postgres server to the docker image. The in-built database server starts only if the environment variable `DJANGO_DATABASE_URL` is not set, this is done to reduce unnecessary resources. 

The following behavior takes place when `DJANGO_DATABASE_URL` env variable is not set
- The inbuilt Postgres server is started
- Default DB user is created with 
  -  username  as `mathesar` 
  -  password  as `mathesar` 
- Default database is created with
  -  database name as `mathesar_django`
- The  `DJANGO_DATABASE_URL` env variable is set to point to the default database running inside the docker image
   - The connection string value for the env variable will be `postgres://mathesar:mathesar@localhost:5432/mathesar_django`
- The  value of the  `MATHESAR_DATABASES` env variable is set to `(mathesar_tables|postgresql://mathesar:mathesar@localhost:5432/mathesar)`

**Technical details**
Most of the information is the same as the [official Postgres Dockerfile](https://github.com/docker-library/postgres/blob/43d17d5ced92f230fa8c196e746f2e2aa288e5e8/13/bullseye/Dockerfile). I made the following changes
- Added the Postgres repo for installing Postgres 13. Debian buster(used by our docker image) comes with Postgres 11 by default
- Set the following default credentials for the database that is created initially
  -  username  as `mathesar` 
  -  password  as `mathesar` 
  -  database as `mathesar_django`

  

## Checklist
 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `develop` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [ ] I added tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: Dockerfile
Comment (mathemancer) [reply Dockerfile:38]: Also the certificates and wget are already installed, I believe. Alsoalso, I don't think we need `wget`. Just use curl.

File: Dockerfile
Comment (mathemancer) [inline Dockerfile:43]: We're not, right? We're starting from the python buster image, which isn't on slim.

File: Dockerfile
Comment (mathemancer) [inline Dockerfile:60]: Why specify this? What's wrong with the defaults?

File: Dockerfile
Comment (mathemancer) [inline Dockerfile:80]: Why does this part exist? The comment doesn't help me understand.

File: Dockerfile
Comment (mathemancer) [inline Dockerfile:81]: Why?? All these things are set up by installing postgres, right?

**Review by mathemancer — CHANGES_REQUESTED** [review]
My main overall comment is that this is _waay_ more complicated than it warrants. I can't see any advantages we're getting over just 
- adding the postgres repo
- Installing postgres
- Running it with `pg_ctl`  (and enabling it so it starts whenever you start the container)

We don't need to be able to do 90% of the stuff enabled by the run script, IMO, and using the above would avoid all manner of weird `chmod`ding and `chroot`ing.

I can already tell by looking at this that no one's going to want to maintain it, and we're going to end up with weird cruft, things you cant touch without breaking, etc. And out-of-date versions of packages.

File: Dockerfile
Comment (silentninja) [reply Dockerfile]: I got a package missing error when I tried to install it from the Debian repo and I ended up copying the instructions from the gosu repo. But it was an oversight on my end, I replaced it with installing from debian repo.

**Comment by silentninja** [comment]
Thanks for the review @mathemancer. I removed most of the irrelevant steps. I didn't remove the steps that change the permissions during runtime because we need it for users running docker without root privileges.

**Review by kgodey — APPROVED** [review]
Made some minor updates to docs. @silentninja you forgot to fix the nav and the main page, I updated those too.

I did not review anything except the docs.

@mathemancer please also review the docs.

File: Dockerfile
Comment (mathemancer) [inline Dockerfile:44]: I think you can remove this part of the comment.

File: docs/docs/installation/docker/index.md
Comment (mathemancer) [inline docs/docs/installation/docker/index.md:28]: missing backslash

File: db-run.sh
Comment (mathemancer) [inline db-run.sh:2]: I think you should `set -e` here. There's no point in continuing if this breaks (is there?).

File: db-run.sh
Comment (mathemancer) [inline db-run.sh:9]: This test is defective. 
- Missing a `$`
- `$PG_VERSION` is not set.

Wouldn't the major version suffice anyway?

**Review by mathemancer — CHANGES_REQUESTED** [review]
This is a major improvement over the last iteration. Thank you for that!

However, while it works in the docker compose setup, it doesn't work when run as a single image. The problem is in the `db-run.sh` script. It has a couple of typos that make the 'is there already a db?' test defective.

### Docs Review
- I suspect we can lower the required docker version by alot. Maybe not worth worrying about right now
- The uninstalling instructions are out-of-sync for the docker install. If they follow the instructions for installing, they won't have the separate DB to worry about.
- There's a missing backslash. See my specific comment.
- We need to add instructions for starting/stopping the service.

**Comment by silentninja** [comment]
@mathemancer I ran across a lot of bugs  or unintented effects trying to use `pg_createcluster`. These are the options I tried

#### pg_createcluster (Works with some inconvenience to the user)

When creating a cluster with `pg_createcluster`, it ends up creating the conf directory in `/etc/postgresql/<version>/<cluster-name>/` instead of the data directory which is in `/var/lib/postgresql/<cluster-name>`. The config files are needed for starting the cluster.  So now the user has to mount `/etc/postgresql/<pg-version>/<cluster-name>/` in addition to the data directory, So in case of a postgres version update the user needs to make changes to the mounted volume path.

#### initdb and pg_ctl (Fails due to permission errors)

I tried looking into `initdb` and `pg_ctl` as a alternative. Both of these tools cannot be run as root and docker initializes a volume folder with root privileges(pg_createcluster changes the ownership during the runtime as it is ran as root process). So `initdb` does cannot change the privilege of the data directory(mounted as a volume) and we need to change the privileges before using `initdb` as a postgres user which is what the code in the initial commit of the PR was actually doing.
 
 
 #### Mix of both pg_createcluster and pg_ctl(Fails on restart)
 
I thought I finally got something working with a good UX in this commit but unfortunately it has its own set of problems too. when the postgres server it always starts in recovery mode. @mathemancer Is there anything I am missing here

**Comment by kgodey** [comment]
> I suspect we can lower the required docker version by alot. Maybe not worth worrying about right now

Please create an issue to track this.

File: db-run.sh
Comment (silentninja) [reply db-run.sh:9]: No, the test is correct. `PG_VERSION` is a file inside the data directory which is created when the cluster created.

File: Dockerfile
Comment (mathemancer) [inline Dockerfile:22]:
```suggestion
VOLUME /etc/postgresql/15/mathesar
```
Also, why not mount `/etc/postgresql/` instead? I thought you were going to try that so we can avoid fiddling around with mounting different version numbers if we want to upgrade and so on.

**Review by mathemancer — CHANGES_REQUESTED** [review]
I think this simplifies things nicely, but:
- there are a number of unaddressed comments from my previous review; some are important.
- I think we need to block this on compatibility with PostgreSQL 15. There's (at least one) issue that keeps the tests from passing on PG15.
- See my specific comment about the mount point

**Comment by silentninja** [comment]
@mathemancer I made changes to Github action to use the docker image and pass the credentials as an environment variable when calling in docker-exec (to avoid breaking any workflow using docker-compose). So we are using the inbuilt postgres server to test Postgres version 13,14 and 15

**Comment by silentninja** [comment]
@mathemancer I addressed all the review feedback, and added the docker image to our test suite. The docs don't need to be updated as there aren't any breaking changes to our test runner. 
I would like to get a final review to make sure your concerns are addressed.

**Review by mathemancer — COMMENTED** [review]
Overall, I think this is getting much better! However:
- Running the container twice results in an error:
```
Starting inbuilt database
Error: cluster configuration already exists
Starting inbuilt database
Error: cluster configuration already exists
Starting inbuilt database
Error: cluster configuration already exists
Starting inbuilt database
Error: cluster configuration already exists
Starting inbuilt database
Error: cluster configuration already exists
Starting inbuilt database
Error: cluster configuration already exists
Starting inbuilt database
Error: cluster configuration already exists
```
See my specific  comment about the documented `run` command.

File: docs/docs/installation/docker/index.md
Comment (mathemancer) [inline docs/docs/installation/docker/index.md:32]: Is this correct? Am I missing something? I think this is misaligned with the volume specified in the dockerfile.

**Review by mathemancer — CHANGES_REQUESTED** [review]
Sorry, mis-clicked on my previous review. It should have been 'request changes' and included the specific comment about the volume mountpoint.

**Review by mathemancer — CHANGES_REQUESTED** [review]
I think the config volume still doesn't match.

**Review by mathemancer — APPROVED** [review]
Okay, I think this is working, and I can find no other problems.

This is a real improvement to the installation for a user trying out Mathesar, nice work!

**Pull request merged**
